pkg/machines/components/vm/bootOrderModal.jsx

```diff
@@ -0,0 +1,319 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+import cockpit from 'cockpit';
+import {
+    Button,
+    ButtonGroup,
+    Icon,
+    ListView,
+    ListViewItem,
+    Modal
+} from 'patternfly-react';
+
+import { ModalError } from './../notification/inlineNotification.jsx';
+import {
+    findHostNodeDevice,
+    getSortedBootOrderDevices,
+    rephraseUI,
+    vmId
+} from '../../helpers.js';
+import {
+    changeBootOrder,
+    getVm
+} from '../../actions/provider-actions.js';
+
+import './bootOrderModal.css';
+
+const _ = cockpit.gettext;
+const UP = -1;
+const DOWN = 1;
+
+function getBootOrderDevices(vm) {
+    let devices = getSortedBootOrderDevices(vm.inactiveXML);
```

Comment:
Well, the thing here is, when the machine is running and you are adjusting boot order and save it, you are adjusting inactive (offline) XML. So it makes sense to list devices from inactive XML.

Now imagine that the machine is running, and somebody edits boot order from outside cockpit (thru virt-manager or virsh or whatever). Because the machine is running, it won't propagate to live XML yet, but it will be in inactive XML, so we want to show to user what's the state in inactive XML.
Otherwise we will rewrite the boot order changes which were done from outside cockpit or it may cause failure from libvirt (e. g. two devices were assigned same value). or (if I change how we update xml for boot order) user after restart gets boot order which he didn't want.

But it's true that this way, we exclude non-permanent disks. Do you have any idea how to solve both of these problems at once?

Reply:
Since we are updating only the offline XML non permanent disks will not exist there so no need to handle them. Thanks for clarification.

However, the fact that we are updating the offline XML only it's our own implementation constraint. If we used https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainUpdateDeviceFlags API instead of redefining the whole domain XML that would be possible.
And to be honest it's nicer to allow people to just change boot order -> reboot and get their changes than having them to shut off the VM completely.

Interaction with other tools like virt-manager can cause inconsistent state to the VM indeed but I would not really put so much effort to avoid such scenarios where some API call fails because the user was messing up with it from two different tools.

So all in all, I am ok with this getting inside as is. But I would like to perform updates to the live XML as well, when there is API for it, like in this case.

Reply:
We cannot do updates to the live XML because you cannot change boot order for running guest.  If you want to add new disk or any other bootable device and you want to boot from it you need to shutdown the guest completely.

Reply:
Okey, so if we cannot change boot order in live config, it resolves this conversation. We can update only offline config.